src/main/java/hudson/tasks/MailSender.java

```diff
@@ -401,66 +377,80 @@ private MimeMessage createEmptyMail(Run<?, ?> build, TaskListener listener) thro
             if(debug)
                 listener.getLogger().println("Trying to send e-mails to individuals who broke the build. sizeof(culprits)=="+culprits.size());
 
-            rcp.addAll(buildCulpritList(listener,culprits));
+            messageBuilder.addRecipients(getUserEmailList(listener, culprits));
         }
         
         // set recipients after filtering out recipients that should not receive emails
-        Set<InternetAddress> filteredRecipients = MailAddressFilter.filterRecipients(build, listener, rcp);
-        if (!filteredRecipients.isEmpty()) {
-            msg.setRecipients(Message.RecipientType.TO,
-                    filteredRecipients.toArray(new InternetAddress[filteredRecipients.size()]));
-        }
+        messageBuilder.setRecipientFilter(new MimeMessageBuilder.AddressFilter() {
+            @Override
+            public Set<InternetAddress> apply(Set<InternetAddress> recipients) {
+                return MailAddressFilter.filterRecipients(build, listener, recipients);
+            }
+        });
```

Comment:
I really think this is a case of "6 of one, half dozen of the other". I thought `toNormalizedAddress` was fine in the original (maybe make it a static util method), but I also think `setRecipientFilter` is fine too and some other API client code was simplified by the hiding of `toNormalizedAddress` and the explicit `addRecipients` methods.

Unless you feel really strongly about it I'd be for leaving as is.